### PR TITLE
[explorer] feat: added copy button beside address

### PR DESCRIPTION
**Pull request by AnthonyLaw**
## What was the issue?
- we like to add a copy button beside the address in the listing table.

## What's the fix?
- added `copyButton` component.
- added `addressField` component, which contains the address and a copy button.
- the copy button will be in  pages `/accounts`, `/transactions`, '/namespaces`, `/mosaics` 

Example:

image

### CODE REVIEW

File: src/components/fields/AddressField.vue [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+				<Truncate>{{value}}</Truncate>
+			</router-link>
+		</div>
+
+		<span>
+			<ButtonCopy
+				title="copy"
+				:value="value"
+				v-if="value"
+				successMessage="addressBeenCopied" />
```
Comment (Jaguar0625) [inline src/components/fields/AddressField.vue:14]: addressHasBeenCopied

File: src/components/controls/ButtonCopy.vue [hunk 2]
```diff
@@ -0,0 +1,64 @@
+<template>
```
Comment (Jaguar0625) [inline src/components/controls/ButtonCopy.vue:1]: unsure of naming conventions, but seems like CopyButton would be a more natural name

File: src/components/controls/ButtonCopy.vue [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	<img
+        :src="IconCopy"
+        class="icon-copy noselect"
+        @click="onCopyClick" />
+</template>
+<script>
```
Comment (Jaguar0625) [inline src/components/controls/ButtonCopy.vue:7]: please watch out for mixed tabs and spaces

File: __tests__/components/controls/ButtonCopy.spec.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+				toast: jest.fn()
+			}
+		}
+	});
+};
+
+const localVue = createLocalVue();
+localVue.use(Vuex);
+
+describe('Create ButtonCopy component', () => {
```
Comment (Jaguar0625) [inline __tests__/components/controls/ButtonCopy.spec.js:40]: ~~Create~~ ButtonCopy component

File: __tests__/components/controls/ButtonCopy.spec.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	const wrapper = setupStoreMount();
+
+	it('processes props data', ()=> {
+		// Assert:
+		expect(wrapper.vm.value).toMatch('copy text');
+		expect(wrapper.vm.successMessage).toMatch('success');
+	});
+
+	it('click on copy button', () =>{
+		// Arrange
```
Comment (Jaguar0625) [inline __tests__/components/controls/ButtonCopy.spec.js:51]: Arrange:

File: __tests__/components/controls/ButtonCopy.spec.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+		document.execCommand = jest.fn();
+
+		// Act:
+		wrapper.vm.onCopyClick();
+
+		// Assert:
+		expect(document.execCommand).toHaveBeenCalledWith('copy');
```
Comment (Jaguar0625) [inline __tests__/components/controls/ButtonCopy.spec.js:58]: try to test toast messages. also try to add test for error case.

File: __tests__/components/fileds/AddressField.spec.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		stubs: {
+			RouterLink: RouterLinkStub
+		}
+	});
+};
+
+const localVue = createLocalVue();
+localVue.use(Vuex);
+
+describe('Create AddressField component', () => {
```
Comment (Jaguar0625) [inline __tests__/components/fileds/AddressField.spec.js:43]: ~~Create~~ AddressField component

File: __tests__/components/fileds/AddressField.spec.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	// Arrange + Act:
+	const wrapper = setupStoreMount();
+
+	it('processes props data', ()=> {
+		// Assert:
+		expect(wrapper.vm.itemKey).toMatch('accounts');
+		expect(wrapper.vm.value).toMatch('NAW7L44MVKCVBM6IGEBXLF2K7JYKEP6R5XMCEZA');
+	});
+
+	it('contain address href', () => {
```
Comment (Jaguar0625) [inline __tests__/components/fileds/AddressField.spec.js:53]: contain => contains (x2)

File: __tests__/components/controls/ButtonCopy.spec.js (hunk 6 above)
Comment (AnthonyLaw) [reply __tests__/components/controls/ButtonCopy.spec.js:58]: Yeah, I did try that, but it not working.
I will try again 👍🏼

**Review by gimre-xymcity — DISMISSED** [review]

File: __tests__/components/controls/ButtonCopy.spec.js (hunk 6 above)
Comment (AnthonyLaw) [reply __tests__/components/controls/ButtonCopy.spec.js:63]: I added unit test individual for toast `errorMsg` and `successMsg` method.

try to work on this, but no luck, not sure I'm in the right direction here.
I mocked `copyTextToClipboard` to assert the function is called, but not working.

still look around how to solve this.

```js
onCopyClick () {
	helper.copyTextToClipboard(this.value) // Mock this
		.then(() => this.successMsg()) // fail assert function is called
		.catch(() => this.errorMsg()); // fail assert function is called
},
```

Comment (AnthonyLaw) [reply __tests__/components/controls/ButtonCopy.spec.js:63]: I found the problem now, all good :)

**Review by Jaguar0625 — APPROVED** [review]

File: src/components/controls/CopyButton.vue [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		value: {
+			type: String,
+			required: true
+		},
+        successMessage: {
+            type: String,
+            default: 'value'
+        }
+	},
+	methods: {
```
[the remark this answers is not in the document]
Comment (AnthonyLaw) [reply src/components/controls/CopyButton.vue]: I fixed that in the latest commit.

**Review by Jaguar0625 — APPROVED** [review]

**Comment by AnthonyLaw** [comment]
@Jaguar0625 Jenkins fail, caused by CSS stylelint.
stylelint (css) used `space` instead of `tab`.

I will create a separate PR for the lint problem.

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**